### PR TITLE
[Fix] Llama 3.2 Vision decoder_trainable flag fixed

### PR DESCRIPTION
**Pull request by pbontrager**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

decoder_trainable flag was not hooked up right. This PR makes it so that setting it to lora or frozen toggles the decoder between frozen weights and lora correctly. This also disables mixed lora and full finetuning training. Setting x_trainable=full for a lora model will now throw an error. This will be re-enabled once checkpointing can properly support this.

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

`tune run lora_finetune_single_device --config llama3_2_vision/11B_lora_single_device metric_logger=torchtune.training.metric_logging.WandBLogger model.decoder_trainable=frozen`

Screenshot 2024-12-11 at 11 08 03 AM

`tune run lora_finetune_single_device --config llama3_2_vision/11B_lora_single_device metric_logger=torchtune.training.metric_logging.WandBLogger model.decoder_trainable=lora`

Screenshot 2024-12-11 at 11 21 35 AM

For further testing I inspected the model parameters with the two options and manually verified that the correct decoder parameters existed or not and were trainable or not.

### CODE REVIEW

File: torchtune/models/llama3_2_vision/_model_builders.py [hunk 1]
```diff
@@ -167,6 +167,11 @@ def lora_llama3_2_vision_11b(
     decoder_type = LoRATrainable(decoder_trainable.lower())
     encoder_type = LoRATrainable(encoder_trainable.lower())
     fusion_type = LoRATrainable(fusion_trainable.lower())
+    assert LoRATrainable.FULL not in [
+        decoder_type,
+        decoder_type,
```
Comment (ebsmothers) [inline torchtune/models/llama3_2_vision/_model_builders.py:172]:
```suggestion
        encoder_type,
```

File: torchtune/models/llama3_2_vision/_model_builders.py [hunk 2]
```diff
@@ -325,6 +330,11 @@ def lora_llama3_2_vision_90b(
...
```
Comment (ebsmothers) [inline torchtune/models/llama3_2_vision/_model_builders.py:335]:
```suggestion
        encoder_type,
```

File: torchtune/models/llama3_2_vision/_model_builders.py [hunk 3]
```diff
@@ -167,6 +167,11 @@ def lora_llama3_2_vision_11b(
...
+        fusion_type,
+    ], "We've temporarily removed support for mixed LoRA + Full Finetuning yet. Please don't use the 'full' option and use llama3_2_vision_11b if you need full finetuning"
```
Comment (ebsmothers) [inline torchtune/models/llama3_2_vision/_model_builders.py:174]: I would just say "X is not currently supported" instead of "We've temporarily removed support for X" (and same comment on the error message further down)

**Review by ebsmothers — COMMENTED** [review]
> For further testing I inspected the model parameters with the two options and manually verified that the correct decoder parameters existed or not and were trainable or not.

Sounds worthy of a unit test to me 👀

File: torchtune/models/llama3_2_vision/_model_builders.py (hunk 3 above)
Comment (pbontrager) [reply torchtune/models/llama3_2_vision/_model_builders.py:174]: I worded it this way since the option is still there, but we don't let you use it. Also in case someone is already using it.

**Comment by pbontrager** [comment]
Unit test added

**Comment by felipemello1** [comment]
> Setting x_trainable=full for a lora model will now throw an error. This will be re-enabled once checkpointing can properly support this.

Just to confirm, it does support it, we just don't support resuming from ckpt. Is that correct?

**Comment by felipemello1** [comment]
For `model.decoder_trainable=True`, should we test it and confirm that it still works? I have yet to look into the code. Did you compare the memory between these 2 wandb runs?

**Review by felipemello1 — DISMISSED** [review]

File: torchtune/models/llama3_2_vision/_model_builders.py (hunk 3 above)
Comment (ebsmothers) [reply torchtune/models/llama3_2_vision/_model_builders.py:174]: > We've temporarily removed support for mixed LoRA + Full Finetuning yet.

😕

**Review by ebsmothers — APPROVED** [review]
Thanks for adding the unit test! I still would update the error message (it reads like a mishmash of two competing sentences). Otherwise this looks good to me

File: torchtune/models/llama3_2_vision/_model_builders.py (hunk 3 above)
Comment (joecummings) [reply torchtune/models/llama3_2_vision/_model_builders.py:174]: Yeah this could be worded in a kinder manner. Feels aggressive. Ask ChatGPT to reword or smthing.

**Review by joecummings — APPROVED** [review]

**Comment by felipemello1** [comment]
i will merge it so i can cut the release. @pbontrager , please write a new PR with a new error msg, and i can cherry pick it
